src/build_packages.sh

```diff
@@ -2,27 +2,54 @@
 set -e
 export XROAD=$(cd "$(dirname "$0")"; pwd)
 
-if [[ ! " $* " =~ " --packageonly " ]]; then
-  ./compile_code.sh "$@"
-fi
+HAS_DOCKER=""
+
+errorExit() {
+	echo "*** $*" 1>&2
+	exit 1
+}
+
+buildInDocker() {
+    test -n "$HAS_DOCKER" || errorExit "Error, docker is not installed/running."
+    echo "Building in docker..."
+    docker build -q -t xroad-build --build-arg uid=$(id -u) --build-arg gid=$(id -g) $XROAD/packages/docker-compile || errorExit "Error building build image."
+    docker run --rm -v $XROAD/..:/workspace -w /workspace/src -u builder xroad-build  bash -c "./update_ruby_dependencies.sh && ./compile_code.sh -nodaemon" || errorExit "Error running build of binaries."
+}
+
+buildLocally() {
+    echo "Building locally..."
+    cd $XROAD || errorExit "Error 'cd $XROAD'."
+    ./compile_code.sh "$@" || errorExit "Error running build of binaries."
+}
 
 if command -v docker &>/dev/null; then
-    docker build -q -t xroad-deb-bionic "$XROAD/packages/docker/deb-bionic"
-    docker build -q -t xroad-deb-focal "$XROAD/packages/docker/deb-focal"
-    docker build -q -t xroad-rpm "$XROAD/packages/docker/rpm"
-    docker build -q -t xroad-rpm-el8 "$XROAD/packages/docker/rpm-el8"
+    HAS_DOCKER=true
+fi
+
+case "$1" in
+    --packageonly|-p) shift;;
+    --docker-build|-d) shift; buildInDocker "$@";;
+    *) buildLocally "$@";;
+esac
```

Comment:
```suggestion
for i in "$@"; do
shift
case "$i" in
    "--package-only"|"-p")
        PACKAGE_ONLY=1
        continue
        ;;
    "--docker-compile"|"-d")
        DOCKER_COMPILE=1
        continue
        ;;
    "--help"|"-h")
         usage
        ;;
    *) set -- "$@" "$i";;
esac
done

if [[ -n "$PACKAGE_ONLY" && -n "$DOCKER_COMPILE" ]]; then
    echo "Can't use both package-only and docker-compile options at the same time"
    usage
fi

if [[ -n "$PACKAGE_ONLY" ]]; then
    echo "Skipping compilation..."
elif [[ -n "$DOCKER_COMPILE" ]]; then
    buildInDocker "$@"
else
    buildLocally "$@"
fi
```

Suggestion: parameter handling in the same fashion as `compile_code.sh`: https://github.com/nordic-institute/X-Road/blob/develop/src/compile_code.sh#L6-L18

- docker-compile, --package-only, -d, -p can appear in any parameter positions, do not have to be the first ones
- those parameters are removed, and retained parameters are carried over to e.g. compile_code.sh
- also minor rewording of parameter names and added -h, --help parameter and usage display

Reply:
I find the construct with `set -- "$@" "$i"` rather confusing. The `continue` statements are unnecessary. 
I really prefer to keep things simple (that's why I personally would be fine with allowing those parameters only in the first position). 

The example from `compile_code.sh` looks better to me because it does not mix a `for` loop with `shift` statements (this should really be avoided). 

If you think a more complete parameter handling is necessary I can dive a bit deeper into it and make a proposal.

Reply:
That set & shift approach is inspired by https://unix.stackexchange.com/a/258514. Not sure why for loop with shift statements should be avoided in general? 

Those `continue`s look unnecessary, I think you are right. 

I did test the proposed solution with different parameter combinations, and all parameters were evaluated if I am not mistaken.

Reply:
I avoid using `shift` in `for` loops altogether because it is very confusing. Shell code needs to be treated with care anyway because scripts are not parsed but rather simply executed line by line (and hence tend to abort in the middle of a file system operation leaving a complete mess). 

That's why I usually recommend to use patterns as clear and simple as possible. When parsing command line args, it's usually either `while` + `shift` or it's `for`. Only in this case it's getting a bit inconvenient because we want to pass on command line args preserving white spaces. Rather than trying to build a new argument list I rather try to avoid having to do so in the first place :)

Reply:
Thanks for the updates!

Looks like this still expects build-mode parameter to be in first position. If we were to do so, there is no reason for the for-loop over all parameters. The earlier syntax of

```
case "$1" in
    --packageonly|-p) shift;;
    --docker-build|-d) shift; buildInDocker "$@";;
    *) shift; buildLocally "$@";;
esac
```

would be simpler and better.

**I will try to argue for the freely order build-mode (d/p) parameters**. The build modes operate a bit weirdly now, as there are two modes that trigger a different-from-default build path, and absence of those triggers the default build path. Everything might a bit more straightforward if there was symmetry and buildLocally was also a mandatory parameter (but I am not very convinced that it would be optimal either). Now, putting the build-mode parameter into wrong index by mistake triggers wrong build path which may be an error that is possibly not immediately easy to detect. Especially when combined with how the sister-script `compile_code.sh` behaves (parameters do have free order) this is a recipe for confusion IMO. 

I know the free-order parameters are not perfect either (potential for masking sharing parameters - but shared parameters should be avoided in any case IMO), but this is still the way I would prefer.

As I put in my suggestion in https://github.com/nordic-institute/X-Road/pull/726#discussion_r543144343, requiring build-mode parameters to be the first ones is also an acceptable solution IMO. I would prefer the free-order solution (especially after I've already spent the time to write and test it), but fixed order with improved documentation would be merged as well.

I agree 100% with the point about shell code needing to be treated with care, and to be as clear and simple as possible, I don't think we have a disagreement there. However things like `shift` in `for` loops _is confusing_ or _is not confusing_ is more subjective, and also down to what solutions each developer is used to. IMO using shift in the for-loop is fine here. Maybe some additional comments would make it more clear:

```
# remove all build-mode params (--package-only,-p,--docker-compile,-d), maintain others
for i in "$@"; do
    # remove current parameter i from all params
    shift
    case "$i" in
        --package-only|-p)
            PACKAGE_ONLY=1
            ;;
        --docker-compile|-d)
            DOCKER_COMPILE=1
            ;;
        --help|-h)
            usage 0
            ;;
        *)
            # bring back removed parameter i
            set -- "$@" "$i"
            ;;
    esac
done
```

Reply:
> I agree 100% with the point about shell code needing to be treated with care, and to be as clear and simple as possible, I don't think we have a disagreement there. However things like `shift` in `for` loops _is confusing_ or _is not confusing_ is more subjective, and also down to what solutions each developer is used to.

We can of course happily agree to disagree on that one :)
Just one last remark, why I don't really favour the construction with the `for` loop: Usually `shift` is used when you always look at `$1`, and that's what you'd do in a `while` loop:

```
while [ $# -gt 0 ]; do
    case "$1" in
        *) ...;; 
    esac
    shift
done
```
In that case I consume the very parameter you're looking at. When using a `for` loop, `shift` does not work the same way, because it's always consuming the first parameter regardless which one is in `$i`. These are code blocks that people writing a lot of shell script code usually have in their heads and just put them in when necessary. Of course, in your case you're always consuming before looking, so the behaviour is consistent and defined, but still it takes some time to grasp. 

> Looks like this still expects build-mode parameter to be in first position. If we were to do so, there is no reason for the for-loop over all parameters. The earlier syntax of
> 
> ```
> case "$1" in
>     --packageonly|-p) shift;;
>     --docker-build|-d) shift; buildInDocker "$@";;
>     *) shift; buildLocally "$@";;
> esac
> ```
> 
> would be simpler and better (copy-pasted from old, uses parameter names no longer in use).
> 
> **I will try to argue for the freely ordered build-mode (d/p) parameters**. The build modes operate a bit weirdly now, as there are two modes that trigger a different-from-default build path, and absence of those triggers the default build path. Everything might a bit more straightforward if there was symmetry and buildLocally was also a mandatory parameter (but I am not very convinced that it would be optimal either). Now, putting the build-mode parameter into wrong index by mistake triggers wrong build path which may be an error that is possibly not immediately easy to detect. Especially when combined with how the sister-script `compile_code.sh` behaves (parameters do have free order) this is a recipe for confusion IMO.

Actually I do not think that the parameters to `build_package.sh` should be freely placable for two reasons:
1. The same parameters (e.g. `-p` or `-d`) could be legal for the script called afterwards. They'd get filtered out here which could cause problems.
2. For me as a user a clear distinction between parameters to the outer and parameters to the inner script makes sense. The simple method is just enforce parameters to the outer script at position 1, a more sophisticated would be to introduce a parameter `--` after which everything will be ignored by the outer script.

For a build script as simple as this one I would prefer not to use the `--` variant, because it makes the script longer than necessary and/or forces me to do "dirty" things like rebuilding `$@`. My original version's charm was its simplicity, but I accept that error handling (`-p` and `-d`) was missing. 

In general I try to keep shell scripts short. My personal convention is to scrutinise everything significantly longer than an editor page and get rid of unnecessary stuff if possible (more code - more bugs, fatal in a programming language without a parser). 

To be honest, I still like my first version best, but the second would still be OK. I don't like the `set -- "$@" "$i"` construct, because basically everybody needs to look up what it does (that's by the way also my standard criticism on `bash`isms), and few people will be able to understand _why_ it works. 

But long story short (I do enjoy this kind of debates): if you really want the variant you proposed I'd be - reluctantly - willing to go along. I'd still prefer the simple one :)

Reply:
Heh, looks like we have created quite a long debate about this detail :) Thanks for explaining your thinking, appreciated. I guess we understand the related factors mostly the same way, but maybe put different weights to different factors when summing up the overall preferred approach.

To recap, I think both options are fine for this PR

1. build-mode params (--package-only,-p,--docker-compile,-d) in fixed first position - close to your 1st version
2. build-mode params in any positions, will get removed for inner script - close to what I proposed

For (1) I would like to see a bit of added documentation, making this as clear as possible for the user, so that she does not `build_packages.sh param1 --docker-build param2` by accident. Usage() at least could use couple of additional words about it, maybe also BUILD.md?

I still prefer (2) but it is not a must.